File: day04/.py
def decode_santa_pin(code: str) -> str | None:
  blocks = code[1:-1].split("][")
    
  if len(blocks) != 4:
    return None
    
  result = ""
    
  for hole in blocks:
    if hole == "<":
      if len(result) == 0:
        return None
      result += result[-1]
      continue
        
    total = int(hole[0])

    for i in range(len(hole)):
    char = hole[i]
      if char == "+":
        total = (total + 1) % 10
      elif char == "-":
        total = (total + 9) % 10 
        
  result += str(total)
    
  return result

File: day04/.ts
/**
 * Deciphers a Santa-style PIN code.
 * @param {string} code - The code to decipher, e.g., "[1+][2-][3][<]"
 * @returns {string|null} The deciphered PIN or null if invalid
 */
function decodeSantaPin(code) {
  let blocks = code.slice(1, -1).split("][");
  
  if (blocks.length !== 4) return null;

  let result = "";

  for (let hole of blocks) {
    if (hole === "<") {
      result += result.at(-1);
      continue;
    }

    let total = Number(hole[0]);
    for (let i = 1; i < hole.length; i++) {
      const char = hole[i];
      if (char === "+") total = (total + 1) % 10;
      else if (char === "-") total = (total - 1 + 10) % 10;
    }

    result += total;
  }

  return result;
}